crates/shared/src/current_block.rs

```diff
@@ -114,12 +113,53 @@ where
             .context("failed to get current block number")?
             .as_u64())
     }
+
+    /// get latest `length` blocks
+    /// if successful, function guarantees `length` blocks in Result (does not return partial results)
+    /// `last_block` block is at the end of the resulting vector
+    async fn preceding_blocks(&self, last_block: u64, length: u64) -> Result<Vec<BlockNumberHash>> {
```

Comment:
Instead of last_block and length we could take one of the range types like std::ops::Range .

Reply:
Yeah. Code looks cleaner now.